### PR TITLE
flavors: Introduce Virtualmachineflavorspec

### PR DESCRIPTION
**Pull request by lyarwood**
**What this PR does / why we need it**:

This change introduces a new shared spec type for the VirtualMachineFlavor CRDs. This drastically simplifies the code required to find and apply these CRDs to a given VM.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

N/A

**Special notes for your reviewer**:

As this API is still in alpha I'm not including a releasenote or docs. Further changes on top of this will however do so when things have stabilised. 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/flavor/flavor.go [hunk 1]
```diff
@@ -15,8 +14,10 @@ import (
 )
 
 type Methods interface {
-	FindProfile(vm *virtv1.VirtualMachine) (*flavorv1alpha1.VirtualMachineFlavorProfile, error)
-	ApplyToVmi(field *k8sfield.Path, profile *flavorv1alpha1.VirtualMachineFlavorProfile, vmiSpec *virtv1.VirtualMachineInstanceSpec) Conflicts
+	FindAndApplyFlavor(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) (Conflicts, error)
```
Comment (akrejcir) [inline pkg/flavor/flavor.go:17]: Why did you combine finding the flavor spec and applying it?
It was simpler to test when they were separate.

File: pkg/flavor/flavor.go [hunk 2]
```diff
@@ -15,8 +14,10 @@ import (
...
+	FindFlavor(vm *virtv1.VirtualMachine) (*flavorv1alpha1.VirtualMachineFlavor, error)
+	FindClusterFlavor(vm *virtv1.VirtualMachine) (*flavorv1alpha1.VirtualMachineClusterFlavor, error)
```
Comment (akrejcir) [inline pkg/flavor/flavor.go:19]: These 2 methods `FindFlavor` and `FindClusterFlavor` are not used anywhere (only in unit tests). Do you plan to use them in a future PR?

File: pkg/flavor/flavor.go [hunk 3]
```diff
@@ -43,95 +44,116 @@ func NewMethods(flavorStore, clusterFlavorStore cache.Store) Methods {
...
 }
 
-func findFirstProfile(profiles []flavorv1alpha1.VirtualMachineFlavorProfile, predicate func(profile *flavorv1alpha1.VirtualMachineFlavorProfile) bool) *flavorv1alpha1.VirtualMachineFlavorProfile {
-	for i := range profiles {
-		profile := &profiles[i]
-		if predicate(profile) {
-			return profile
+func (m *methods) FindFlavorSpec(vm *virtv1.VirtualMachine) (*flavorv1alpha1.VirtualMachineFlavorSpec, error) {
+
+	switch strings.ToLower(vm.Spec.Flavor.Kind) {
```
Comment (akrejcir) [inline pkg/flavor/flavor.go:75]: Can you add a `case` for cluster flavors and `default` branch that returns an error, if the `Kind` does not match any case?

File: pkg/flavor/flavor.go [hunk 4]
```diff
@@ -43,95 +44,116 @@ func NewMethods(flavorStore, clusterFlavorStore cache.Store) Methods {
...
+	if vmi.Annotations == nil {
+		vmi.Annotations = make(map[string]string)
+	}
+
+	switch strings.ToLower(vm.Spec.Flavor.Kind) {
+	case apiflavor.SingularResourceName, apiflavor.PluralResourceName:
+		vmi.Annotations[virtv1.FlavorAnnotation] = vm.Spec.Flavor.Name
+	case apiflavor.ClusterSingularResourceName, apiflavor.ClusterPluralResourceName, "":
+		vmi.Annotations[virtv1.ClusterFlavorAnnotation] = vm.Spec.Flavor.Name
+	}
```
Comment (akrejcir) [inline pkg/flavor/flavor.go:137]: You can call function `AddFlavorNameAnnotations()` here.

File: pkg/virt-api/webhooks/validating-webhook/admitters/flavor-admitter.go [hunk 5]
```diff
@@ -26,42 +25,68 @@ func (f *FlavorAdmitter) Admit(ar *admissionv1.AdmissionReview) *admissionv1.Adm
 			Version:  flavorv1alpha1.SchemeGroupVersion.Version,
 			Resource: flavor.PluralResourceName,
 		},
-		func(raw []byte) ([]flavorv1alpha1.VirtualMachineFlavorProfile, error) {
-			flavorObj := flavorv1alpha1.VirtualMachineFlavor{}
+		func(raw []byte) (*flavorv1alpha1.VirtualMachineFlavor, error) {
```
Comment (akrejcir) [inline pkg/virt-api/webhooks/validating-webhook/admitters/flavor-admitter.go:28]: This function can return `VirtualMachineFlavorSpec` and then the webhook can use common code for Flavor and ClusterFlavor, similarly to what was done with profiles.

**Review by akrejcir — CHANGES_REQUESTED** [review]
Can you squash the two commits? The first commit changes a lot of code that is changed again in the second commit. I think it would be easier to understand.

File: pkg/flavor/flavor.go (hunk 2 above)
Comment (lyarwood) [reply pkg/flavor/flavor.go:19]: I think I did at the time but yeah you're right they don't need to be public, I'll change this.

File: pkg/flavor/flavor.go (hunk 4 above)
Comment (lyarwood) [reply pkg/flavor/flavor.go:137]: Ah bad rebase apologies, I think I've switched over to this later in the series. I'll switch this over here.

File: pkg/flavor/flavor.go (hunk 3 above)
Comment (lyarwood) [reply pkg/flavor/flavor.go:75]: Done.

File: pkg/virt-api/webhooks/validating-webhook/admitters/flavor-admitter.go (hunk 5 above)
Comment (lyarwood) [reply pkg/virt-api/webhooks/validating-webhook/admitters/flavor-admitter.go:28]: Yeah I could do that but I think I wanted to keep a reference to the overall flavorobj on the off chance we need to validate anything outside of the spec in the future. FWIW I add some validation later on using the spec here:

https://github.com/kubevirt/kubevirt/pull/7578/commits/9dd3454726d45fd0ad9e92e7363d4278986f3f18

File: pkg/flavor/flavor.go (hunk 1 above)
Comment (akrejcir) [reply pkg/flavor/flavor.go:18]: I have checked your other PR to see how are these new methods used. For example in `vms-admitter.go` in `applyFlavorToVm()` method: First `FindFlavorSpec(vm)` is called, then `FindPreferenceSpec(vm)` and then later `FindAndApplyFlavor(vm, vmi)` which repeats the same calls.

If the apply method is separate, then the find calls would be done only once. Maybe you can add a free-standing helper function to call find followed by apply?

File: pkg/testutils/mock_flavor.go [hunk 6]
```diff
@@ -9,26 +9,24 @@ import (
 )
 
 type MockFlavorMethods struct {
-	FindFlavorFunc func(vm *v1.VirtualMachine) (*flavorv1alpha1.VirtualMachineFlavorProfile, error)
-	ApplyToVmiFunc func(field *k8sfield.Path, profile *flavorv1alpha1.VirtualMachineFlavorProfile, vmiSpec *v1.VirtualMachineInstanceSpec) flavor.Conflicts
+	FindFlavorSpecFunc func(vm *v1.VirtualMachine) (*flavorv1alpha1.VirtualMachineFlavorSpec, error)
+	ApplyToVmiFunc     func(field *k8sfield.Path, flavorspec *flavorv1alpha1.VirtualMachineFlavorSpec, vmiSpec *v1.VirtualMachineInstanceSpec) flavor.Conflicts
 }
 
-var _ flavor.Methods = &MockFlavorMethods{}
```
Comment (akrejcir) [inline pkg/testutils/mock_flavor.go:16]: This is a golang pattern to make the compiler verify that the struct implements the interface.
It is useful when changing methods in the interface, so we don't forget to change the struct that implements it.
The variable is probably optimized out on compilation.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter_test.go [hunk 7]
```diff
@@ -41,8 +41,10 @@ import (
 	k8sfield "k8s.io/apimachinery/pkg/util/validation/field"
 	"k8s.io/client-go/tools/cache"
 
+	apiflavor "kubevirt.io/api/flavor"
+	"kubevirt.io/api/flavor/v1alpha1"
```
Comment (akrejcir) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter_test.go:45]: No need to rename this import.

**Review by akrejcir — COMMENTED** [review]
Looks good. My comments are just nit-picks, you can ignore them if you want.

/lgtm

File: pkg/testutils/mock_flavor.go (hunk 6 above)
Comment (lyarwood) [reply pkg/testutils/mock_flavor.go:16]: ACK.

File: staging/src/kubevirt.io/api/flavor/v1alpha1/types.go [hunk 8]
```diff
@@ -35,9 +35,8 @@ type VirtualMachineFlavor struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 
-	// +listType=map
-	// +listMapKey=name
-	Profiles []VirtualMachineFlavorProfile `json:"profiles"`
+	// VirtualMachineFlavorSpec for the flavor
+	Spec *VirtualMachineFlavorSpec `json:"spec"`
```
Comment (akrejcir) [inline staging/src/kubevirt.io/api/flavor/v1alpha1/types.go:39]: `Spec` should not be a pointer, because it is not an optional filed.

**Review by akrejcir — CHANGES_REQUESTED** [review]
/lgtm cancel

File: staging/src/kubevirt.io/api/flavor/v1alpha1/types.go (hunk 8 above)
Comment (lyarwood) [reply staging/src/kubevirt.io/api/flavor/v1alpha1/types.go:39]: ACK I'll fix this now, thanks!

**Comment by vladikr** [comment]
Looks great!
/approve

**Comment by lyarwood** [comment]
> /hold until the sig-{network,preformance} jobs are fixed

https://github.com/kubevirt/project-infra/pull/2059 *might* help here, the sig-network issue looks intermittent.

**Comment by lyarwood** [comment]
sig-performance is now hitting https://github.com/kubevirt/kubevirt/issues/7658 FWIW.

**Pull request merged**